File: app/lib/core/ble/telemetry_codec.dart
/// Decoder for the 15-byte BLE telemetry struct.
///
/// Authoritative spec: `contracts/ble_gatt.md`
/// Golden vectors:     `contracts/vectors/telemetry.json`
///
/// The Python backend has a mirror of this in
/// `backend/app/services/telemetry_codec.py`. Both suites decode the same
/// vector file, so a wire-format change that lands on only one side fails both
/// builds rather than silently corrupting readings on one platform.
library;

import 'dart:typed_data';

/// Sentinels marking a failed sensor read. Distinct from a real zero — a dead
/// I2C bus must never be indistinguishable from 0.0 °C.
const int kTempFault = -32768;
const int kRhFault = 0xFFFF;
const int kSoilFault = 0xFFFF;
const int kLuxFault = 0xFFFFFFFF;

const int kFlagPumpOn = 1 << 0;
const int kFlagReservoirLow = 1 << 1;
const int kFlagSensorFault = 1 << 2;
const int kFlagUncalibrated = 1 << 3;
const int kFlagWifiConnected = 1 << 4;
const int kFlagMqttConnected = 1 << 5;

const int kTelemetrySize = 15;

class Telemetry {
  const Telemetry({
    required this.uptimeSeconds,
    required this.temperatureC,
    required this.humidity,
    required this.soilPercent,
    required this.lux,
    required this.flags,
  });

  final int uptimeSeconds;

  /// Null when the sensor reported a fault.
  final double? temperatureC;
  final double? humidity;
  final double? soilPercent;
  final double? lux;

  final int flags;

  bool get pumpOn => flags & kFlagPumpOn != 0;
  bool get reservoirLow => flags & kFlagReservoirLow != 0;
  bool get sensorFault => flags & kFlagSensorFault != 0;

  /// Soil readings are not trustworthy until the air/water calibration has
  /// been completed. The UI must prompt rather than display a percentage.
  bool get uncalibrated => flags & kFlagUncalibrated != 0;

  bool get wifiConnected => flags & kFlagWifiConnected != 0;
  bool get mqttConnected => flags & kFlagMqttConnected != 0;

  /// Decodes a notification payload from the `live_telemetry` characteristic.
  ///
  /// Throws [FormatException] on a wrong-sized buffer rather than accepting a
  /// truncated read — a short frame means the contract is broken somewhere,
  /// and guessing at the missing bytes would surface fiction as sensor data.
  factory Telemetry.decode(List<int> raw) {
    if (raw.length != kTelemetrySize) {
      throw FormatException(
        'expected $kTelemetrySize bytes, got ${raw.length}',
      );
    }

    final bytes = raw is Uint8List ? raw : Uint8List.fromList(raw);
    final data = ByteData.sublistView(bytes);

    // Little-endian throughout, matching the ESP32's native layout.
    final uptime = data.getUint32(0, Endian.little);
    final tempRaw = data.getInt16(4, Endian.little);
    final rhRaw = data.getUint16(6, Endian.little);
    final soilRaw = data.getUint16(8, Endian.little);
    final luxRaw = data.getUint32(10, Endian.little);
    final flags = data.getUint8(14);

    return Telemetry(
      uptimeSeconds: uptime,
      temperatureC: tempRaw == kTempFault ? null : tempRaw / 100.0,
      humidity: rhRaw == kRhFault ? null : rhRaw / 100.0,
      soilPercent: soilRaw == kSoilFault ? null : soilRaw / 100.0,
      lux: luxRaw == kLuxFault ? null : luxRaw / 10.0,
      flags: flags,
    );
  }

  /// Inverse of [Telemetry.decode]. Used by tests and the BLE fake.
  Uint8List encode() {
    final data = ByteData(kTelemetrySize);
    data.setUint32(0, uptimeSeconds, Endian.little);
    data.setInt16(
      4,
      temperatureC == null ? kTempFault : (temperatureC! * 100).round(),
      Endian.little,
    );
    data.setUint16(
      6,
      humidity == null ? kRhFault : (humidity! * 100).round(),
      Endian.little,
    );
    data.setUint16(
      8,
      soilPercent == null ? kSoilFault : (soilPercent! * 100).round(),
      Endian.little,
    );
    data.setUint32(
      10,
      lux == null ? kLuxFault : (lux! * 10).round(),
      Endian.little,
    );
    data.setUint8(14, flags & 0xFF);
    return data.buffer.asUint8List();
  }

  @override
  String toString() =>
      'Telemetry(uptime: ${uptimeSeconds}s, temp: $temperatureC, '
      'rh: $humidity, soil: $soilPercent, lux: $lux, '
      'flags: 0x${flags.toRadixString(16).padLeft(2, '0')})';
}

File: backend/app/services/telemetry_codec.py
"""Codec for the 15-byte BLE telemetry struct.

Authoritative spec: contracts/ble_gatt.md
Golden vectors:     contracts/vectors/telemetry.json (shared with the Dart tests)
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, asdict
from typing import Final

STRUCT_FORMAT: Final = "<IhHHIB"
STRUCT_SIZE: Final = struct.calcsize(STRUCT_FORMAT)
assert STRUCT_SIZE == 15, f"contract violation: struct is {STRUCT_SIZE} bytes, must be 15"

# Sentinels marking a failed sensor read. Distinct from a real zero, which is
# why they exist at all: a dead I2C bus must not look like 0.0 C.
TEMP_FAULT: Final = -32768
RH_FAULT: Final = 0xFFFF
SOIL_FAULT: Final = 0xFFFF
LUX_FAULT: Final = 0xFFFFFFFF

FLAG_PUMP_ON: Final = 1 << 0
FLAG_RESERVOIR_LOW: Final = 1 << 1
FLAG_SENSOR_FAULT: Final = 1 << 2
FLAG_UNCALIBRATED: Final = 1 << 3
FLAG_WIFI_CONNECTED: Final = 1 << 4
FLAG_MQTT_CONNECTED: Final = 1 << 5


@dataclass(frozen=True)
class Telemetry:
    uptime_s: int
    temp_c: float | None
    rh: float | None
    soil_pct: float | None
    lux: float | None
    flags: int

    @property
    def pump_on(self) -> bool:
        return bool(self.flags & FLAG_PUMP_ON)

    @property
    def reservoir_low(self) -> bool:
        return bool(self.flags & FLAG_RESERVOIR_LOW)

    @property
    def sensor_fault(self) -> bool:
        return bool(self.flags & FLAG_SENSOR_FAULT)

    @property
    def uncalibrated(self) -> bool:
        return bool(self.flags & FLAG_UNCALIBRATED)

    def to_dict(self) -> dict:
        return asdict(self)


def decode(raw: bytes) -> Telemetry:
    """Decode a BLE notification payload.

    Raises ValueError on a wrong-sized buffer rather than silently accepting a
    truncated read — a short frame means the contract is broken somewhere, and
    guessing at the missing bytes would put fiction into the database.
    """
    if len(raw) != STRUCT_SIZE:
        raise ValueError(f"expected {STRUCT_SIZE} bytes, got {len(raw)}")

    uptime, temp_raw, rh_raw, soil_raw, lux_raw, flags = struct.unpack(STRUCT_FORMAT, raw)

    return Telemetry(
        uptime_s=uptime,
        temp_c=None if temp_raw == TEMP_FAULT else temp_raw / 100.0,
        rh=None if rh_raw == RH_FAULT else rh_raw / 100.0,
        soil_pct=None if soil_raw == SOIL_FAULT else soil_raw / 100.0,
        lux=None if lux_raw == LUX_FAULT else lux_raw / 10.0,
        flags=flags,
    )


def encode(t: Telemetry) -> bytes:
    """Inverse of `decode`. Used by tests and the device simulator."""
    return struct.pack(
        STRUCT_FORMAT,
        t.uptime_s & 0xFFFFFFFF,
        TEMP_FAULT if t.temp_c is None else round(t.temp_c * 100),
        RH_FAULT if t.rh is None else round(t.rh * 100),
        SOIL_FAULT if t.soil_pct is None else round(t.soil_pct * 100),
        LUX_FAULT if t.lux is None else round(t.lux * 10),
        t.flags & 0xFF,
    )


# --- MQTT-side validation -------------------------------------------------

PLAUSIBLE_RANGES: Final[dict[str, tuple[float, float]]] = {
    "temp_c": (-40.0, 85.0),
    "rh": (0.0, 100.0),
    "soil_pct": (0.0, 100.0),
    "lux": (0.0, 200_000.0),
    "batt_mv": (0.0, 6000.0),
}


def clamp_sample(sample: dict) -> tuple[dict, list[str]]:
    """Null out physically impossible values before they reach the database.

    Returns the cleaned sample and the list of fields rejected. A shorted ADC
    pin reading -3000 C should not end up as a data point that flattens every
    temperature chart the user ever looks at.
    """
    cleaned = dict(sample)
    rejected: list[str] = []

    for field, (lo, hi) in PLAUSIBLE_RANGES.items():
        value = cleaned.get(field)
        if value is None:
            continue
        try:
            numeric = float(value)
        except (TypeError, ValueError):
            cleaned[field] = None
            rejected.append(field)
            continue
        if not (lo <= numeric <= hi):
            cleaned[field] = None
            rejected.append(field)

    if rejected:
        cleaned["flags"] = int(cleaned.get("flags") or 0) | FLAG_SENSOR_FAULT

    return cleaned, rejected
